buildSrc/src/main/java/org/elasticsearch/gradle/internal/rest/compat/RestCompatTestTransformTask.java

```diff
@@ -0,0 +1,151 @@
+import com.fasterxml.jackson.databind.SequenceWriter;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import com.fasterxml.jackson.dataformat.yaml.YAMLFactory;
+import com.fasterxml.jackson.dataformat.yaml.YAMLParser;
+import org.elasticsearch.gradle.test.rest.transform.InjectHeaders;
+import org.elasticsearch.gradle.test.rest.transform.RestTestTransform;
+import org.elasticsearch.gradle.test.rest.transform.RestTestTransformer;
+import org.elasticsearch.gradle.util.GradleUtils;
+import org.gradle.api.DefaultTask;
+import org.gradle.api.Project;
+import org.gradle.api.file.FileTree;
+import org.gradle.api.plugins.JavaPluginConvention;
+import org.gradle.api.tasks.Input;
+import org.gradle.api.tasks.InputFiles;
+import org.gradle.api.tasks.OutputDirectory;
+import org.gradle.api.tasks.SkipWhenEmpty;
+import org.gradle.api.tasks.SourceSet;
+import org.gradle.api.tasks.TaskAction;
+import org.gradle.api.tasks.util.PatternFilterable;
+import org.gradle.api.tasks.util.PatternSet;
+import org.gradle.internal.Factory;
+
+import javax.inject.Inject;
+import java.io.File;
+import java.io.IOException;
+import java.util.Collections;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+public class RestCompatTestTransformTask extends DefaultTask {
+
+    private static final YAMLFactory yaml = new YAMLFactory();
+    private static final ObjectMapper mapper = new ObjectMapper(yaml);
+
+    private static final Map<String, String> headers = Map.of(
+        "Content-Type",
+        "application/vnd.elasticsearch+json;compatible-with=7",
+        "Accept",
+        "application/vnd.elasticsearch+json;compatible-with=7"
+    );
+
+    private String sourceSetName;
+    private static final String REST_TEST_PREFIX = "rest-api-spec/test";
+
+    private final PatternFilterable testPatternSet;
+    private final List<RestTestTransform<?>> transformations;
+
+    private String inputResourceParent = "";
+
+    public RestCompatTestTransformTask() {
+        testPatternSet = getPatternSetFactory().create();
+        testPatternSet.include("/*" + "*/*.yml"); // concat these strings to keep build from thinking this is invalid javadoc
+        transformations = Collections.singletonList(new InjectHeaders(headers));
+    }
+
+    @Inject
+    protected Factory<PatternSet> getPatternSetFactory() {
+        throw new UnsupportedOperationException();
+    }
+
+    @Input
+    public String getSourceSetName() {
+        return sourceSetName;
+    }
+
+    @OutputDirectory
+    public File getOutputDir() {
+        return new File(
+            getSourceSet().orElseThrow(() -> new IllegalArgumentException("could not find source set [" + sourceSetName + "]"))
+                .getOutput()
+                .getResourcesDir(),
+            REST_TEST_PREFIX
+        );
+    }
+
+    @SkipWhenEmpty
+    @InputFiles
+    public FileTree getTestFiles() {
+        return getProject().files(
```

Comment:
This is very hard to grok, especially since `inputResourceParent` is actually a "child" directory. I'd much prefer this logic also go into the plugin and the task imlementation being as light as possible. We should give it a `FileCollection` of input and a `FIle` for output directory, that's it. The constuction of those thould be done in the plugin. That has the added benefit that we can completely remove any knowledge of a "source set" from this task altogether. All it does is take a set of input yaml files, do some stuff to them, and dump them in a location which you specify.

Reply:
+1

Reply:
Done.  I left the .yml filter in the task since we have to create YAML specific parser and wanted to keep that association close together.